tensorboard/schema.py

```diff
@@ -0,0 +1,262 @@
+    name='Plugins',
+    columns=[ColumnSchema('plugin_id', Int64ColumnType()),
+             ColumnSchema('name', StringColumnType(length=255), not_null=True)],
+    keys=['plugin_id'])
+
+RUNS_TABLE = TableSchema(
+    name='Runs',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('experiment_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('run_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('name', StringColumnType(length=1900),
+                          not_null=True)],
+    keys=['rowid', 'customer_number', 'experiment_id', 'run_id'])
+
+TAGS_TABLE = TableSchema(
+    name='Tags',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('run_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('tag_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('plugin_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('name', StringColumnType(500)),
+             ColumnSchema('display_name', StringColumnType(500)),
+             ColumnSchema('summary_description', StringColumnType(65535)),],
+    keys=['rowid', 'customer_number', 'run_id', 'tag_id'])
+
+TENSORS_TABLE = TableSchema(
+    name='Tensors',
+    columns=[ColumnSchema('rowid', Int64ColumnType()),
+             ColumnSchema('customer_number', Int64ColumnType()),
+             ColumnSchema('tag_id', Int64ColumnType(), not_null=True),
+             ColumnSchema('step_count', Int64ColumnType(), not_null=True),
+             ColumnSchema('encoding', Int64ColumnType(), not_null=True),
+             ColumnSchema('is_big', BoolColumnType(), not_null=True),
+             ColumnSchema('tensor', BytesColumnType(), not_null=True)],
+    keys=['rowid', 'customer_number', 'tag_id', 'step_count'])
+
+# List of all tables.
+TABLES = [BIG_TENSORS_TABLE, EVENT_LOGS_TABLE, EXPERIMENTS_TABLE, PLUGINS_TABLE,
+          RUNS_TABLE, TAGS_TABLE, TENSORS_TABLE]
+
+
+class IndexSchema(object):
+  """Define the schema for an index."""
+
+  def __init__(self, name, table, columns):
+    """Create the schema for an index.
+
+    Args:
+      name: Name for the index.
+      table: Name of the table to create the index from.
+      columns: The names of the columns comprising the index.
+
+    Returns:
+      IndexSchema representing the schema.
+
+    : type name: str
+    : type table: str
+    : type columns: list[str]
+    : rtype: IndexSchema
+    """
+
+    self.name = name
+    self.table = table
+    self.columns = columns
+
+
+EXPERIMENTS_NAME_INDEX = IndexSchema('ExperimentsNameIndex', 'Experiments', [
+    'customer_number', 'name'])
+
+EVENT_LOGS_PATH_INDEX = IndexSchema('EventLogsPathIndex', 'EventLogs', [
+    'customer_number', 'run_id', 'path'])
+
+PLUGINS_NAME_INDEX = IndexSchema('PluginsNameIndex', 'Plugins', ['name'])
```

Comment:
One concern I have about this design is that the column names are strings that are defined multiple times. I understand that's a problem with the existing hard-coded SQL. The only reason I have this concern is because I feel, if a higher level definition-oriented abstraction is to be offered, then maybe this type of redundancy is one of the problems it should solve.

For example, if we took a metaclass approach to defining schemas, like Django, this could be written:

```py
class Plugins(Table):
  plugin_id = Int64ColumnType()
  name = StringColumnType(length=255, not_null=True)

PLUGINS_NAME_INDEX = Index('PluginsNameIndex', Plugins, Plugins.name)
```

Or it could be done more arduously with the current model:

```py
PLUGINS_NAME_INDEX = Index('PluginsNameIndex', PLUGINS_TABLE_NAME, PLUGINS_TABLE_NAME_COLUMN_NAME)
```

Reply:
Per our IM chat. I punted. So current plan is to just manually maintain separate schemas for different data types.